### PR TITLE
Add support to record `Hint` in `metrics.go` for instant and range queries

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
This PR adds support to record `hint` from the instant and range queries in `metrics.go`. 
Rationale being, going forward we use `hint` as way to test some features under feature toggle (say `Full Log Volume` query currently have `logvolhist` as `hint`.

This will print additional `hint` in the `metrics.go` log line.
e.g:
```
"level=info org_id=foo traceID=%s latency=slow query=\"{foo=\\\"bar\\\"} |= \\\"buzz\\\"\" query_type=filter range_type=range length=1h0m0s step=1m0s duration=25.25s status=200 limit=1000 hint=logvolhist returned_lines=10 throughput=100kB total_bytes=100kB\n",

```

**Which issue(s) this PR fixes**:
Fixes # NA

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/logcli/client/file.go
Comment (dannykopping) [inline pkg/logcli/client/file.go:81]: What do you thinking about hinting `logcli` here? Do we already capture the user agent somewhere perhaps?

File: pkg/loghttp/params.go
Comment (dannykopping) [inline pkg/loghttp/params.go:79]: We should be careful here about using unsanitized user input; this is a common attack vector.

File: pkg/ruler/compat.go
Comment (dannykopping) [inline pkg/ruler/compat.go:73]: Ditto; can we add a `ruler` hint?

**Review by dannykopping — COMMENTED** [review]
LGTM, but I have a security concern

File: pkg/logcli/client/file.go
Comment (kavirajk) [reply pkg/logcli/client/file.go:81]: This empty string is just to satisfy the interface.(And at the moment, I don't know if we need to accept explicity `hint` for query from LogCLI)

And we do capture User agent for logcli here
https://github.com/grafana/loki/blob/main/pkg/logcli/client/client.go#L181

Although I don't know how is it related?

File: pkg/ruler/compat.go
Comment (kavirajk) [reply pkg/ruler/compat.go:73]: I didn't understand. what is `ruler` hint?

File: pkg/logcli/client/file.go
Comment (dannykopping) [reply pkg/logcli/client/file.go:81]: What I'm getting at is: should we use this `hint` to also inform us when different clients are used? This might come in handy one day. If we're already logging the User-Agent then we probably don't need to.

File: pkg/ruler/compat.go
Comment (dannykopping) [reply pkg/ruler/compat.go:73]: This is code in the `ruler`.

File: pkg/logcli/client/file.go
Comment (kavirajk) [reply pkg/logcli/client/file.go:81]: Ah I see. So currently we plan to use `hint` for mainly feature toggles. But sure yea, it could have other use cases going forward.

And yea, we may not need to use it for User-Agent, as we already set that header for any request sent by LogCLI.

File: pkg/loghttp/params.go
Comment (kavirajk) [reply pkg/loghttp/params.go]: Good point. Now we allow only alpha-numeric and `-` characters in the `hint` string!

Comment (dannykopping) [reply pkg/loghttp/params.go]: Thanks!

File: pkg/logcli/client/file.go
Comment (dannykopping) [reply pkg/logcli/client/file.go:81]: Sure, but do we _log_ the user-agent for these requests?
Don't let this comment block the PR, but I think it might be worth considering this

**Review by dannykopping — APPROVED** [review]
LGTM!
Please wait for @slim-bean to approve before merging, since he was more involved in this

**Comment by slim-bean** [comment]
Cyril pinged me to ask me about this and raised a good question, if perhaps `hint` isn't the best term here. Traditionally hints are passed in queries to be used for optimizations on query engines, which is actually what I was thinking when using hint here. But there may be more appropriate uses of hint inside our query engine.

I wonder if instead we should consider this a `type` of query and introduce it as an alternative to `instant` or `range`? I wonder if this is also a bad idea to overload and if we should consider another option of another descriptor all together???

Ultimately I think there is value in having Loki know this is a log volume histogram query so we can make decisions about it (set a timeout, control parallelism etc) and that we should indicate this in `metrics.go` so we can do some analysis on this type of query.

**Review by sandeepsukhani — APPROVED** [review]
LGTM

**Review by jeschkies — APPROVED** [review]
I'm not really sure I understand what `hint` is. Is this to ease debugging or is this rather a feature flag that I can pass with a query?

**Comment by kavirajk** [comment]
Closing this in correspondence to #4814

**Pull request closed without merging**
